File: tests/mytrees.erl
-module(mytrees).
-export([tree1/1, tree2/1, tree3/1, tree4/1, tree5/1, tree/1]).
-export([member/2, join/2, delete/2, tree_sum/1, preorder/1]).

%% NOTE: Must compile with debug_info enabled.

-include_lib("proper/include/proper.hrl").

%==============================================================================

%% API for trees

-type tree(T) :: 'leaf' | {'node',T,tree(T),tree(T)}.

-spec member(T, tree(T)) -> boolean().
member(_, leaf) ->
    false;
member(X, {node,X,_,_}) ->
    true;
member(X, {node,_,L,R}) ->
    member(X, L) orelse member(X, R).

-spec join(tree(T), tree(T)) -> tree(T).
join(leaf, T2) ->
    T2;
join({node,X1,L1,R1}, T2) ->
    {node,X1,join(L1, R1),T2}.

%% Bad implementation of delete: doesn't delete any occurence of X that is
%% nested under some other node containing X.
-spec delete(T, tree(T)) -> tree(T).
delete(_, leaf) ->
    leaf;
delete(X, {node,X,L,R}) ->
    join(L, R);
delete(X, {node,Y,L,R}) ->
    {node,Y,delete(X, L),delete(X, R)}.

-spec tree_sum(tree(number())) -> number().
tree_sum(leaf) ->
    0;
tree_sum({node,X,L,R}) ->
    tree_sum(L) + tree_sum(R) + X.

-spec preorder(tree(T)) -> [T].
preorder(leaf) ->
    [];
preorder({node,X,L,R}) ->
    [X] ++ preorder(L) ++ preorder(R).

%==============================================================================

%% Properties of the tree API

prop_delete() ->
    ?FORALL({X,T}, {integer(),tree(integer())},
	    not member(X, delete(X, T))).

%% Example of a PBT strategy: finding two distinct computations that should
%% result in the same value.
prop_sum() ->
    ?FORALL(T, tree(integer()),
	    lists:sum(preorder(T)) =:= tree_sum(T)).

%==============================================================================

%% Attempts at writing a generator for trees:

tree1(G) ->
    union(['leaf',
	   {'node',G,tree1(G),tree1(G)}
	  ]).

%% Erlang is eager: we need to enforce lazy evaluation to avoid infinite
%% recursion.

tree2(G) ->
    union(['leaf',
	   ?LAZY({'node',G,tree2(G),tree2(G)})
	  ]).

%% Generation might not terminate: we need to introduce a bound on the number
%% of recursive calls (and thus the size of the produced term), by handling the
%% 'size' parameter manually.

%% The base case is delegated to the 0-size clause.
%% All non-recursive cases are replaced by fallbacks to that clause.

tree3(G) ->
    ?SIZED(S, tree3(S, G)).

tree3(0, _) ->
    leaf;
tree3(S, G) ->
    union([tree3(0, G),
	   ?LAZY({'node',G,tree3(S-1, G),tree3(S-1, G)})
	  ]).

%% 50% of the time, the tree is empty: we should set the weights in the union
%% to ensure a satisfactory average size of produced instances.

tree4(G) ->
    ?SIZED(S, tree4(S, G)).

tree4(0, _) ->
    'leaf';
tree4(S, G) ->
    frequency([{1, tree4(0, G)},
	       {9, ?LAZY({'node',G,tree4(S-1, G),tree4(S-1, G)})}
	      ]).

%% The trees grow too fast: we should distribute the size evenly to all
%% subtrees.

tree5(G) ->
    ?SIZED(S, tree5(S, G)).

tree5(0, _) ->
    'leaf';
tree5(S, G) ->
    frequency([{1, tree5(0, G)},
	       {9, ?LAZY({'node',G,tree5(S div 2, G),tree5(S div 2, G)})}
	      ]).

%% Finally, we set up a more efficient shrinking strategy: pick each of the
%% subtrees in place of the tree that fails the property.

tree(G) ->
    ?SIZED(S, tree(S, G)).

tree(0, _) ->
    'leaf';
tree(S, G) ->
    frequency([{1, tree(0, G)},
	       {9, ?LAZY(?LETSHRINK([L,R],
				    [tree(S div 2, G),tree(S div 2, G)],
				    {'node',G,L,R}
				   ))}
	      ]).

%% Try this out using:
%% T2 = mytrees:tree(proper_types:integer()).
%% f(I), {ok,I} = proper_gen:pick(T2), io:format("~p~n",[I]).

%% Comment out the generator and recompile to use the type instead.
%% Try this out using:
%% {ok,T1} = proper_typeserver:demo_translate_type(mytrees, "tree(integer())").
%% f(I), {ok,I} = proper_gen:pick(T1), io:format("~p~n",[I]).

File: tests/mytrees.pl
fundef(lit(atom,'mytrees'),var('member',2),fun([var('@c1'),var('@c0')],case([var('@c1'),var('@c0')],[clause([var('@c6'),lit(atom,'leaf')],lit(atom,'true'),lit(atom,'false')),clause([var('X'),tuple([lit(atom,'node'),var('@c7'),var('@c8'),var('@c9')])],call(lit(atom,'erlang'),lit(atom,'=:='),[var('@c7'),var('X')]),lit(atom,'true')),clause([var('X'),tuple([lit(atom,'node'),var('@c10'),var('L'),var('R')])],lit(atom,'true'),case([apply(var('member',2),[var('X'),var('L')])],[clause([lit(atom,'true')],lit(atom,'true'),lit(atom,'true')),clause([lit(atom,'false')],lit(atom,'true'),apply(var('member',2),[var('X'),var('R')])),clause([var('@c2')],lit(atom,'true'),call(lit(atom,'erlang'),lit(atom,'error'),[tuple([lit(atom,'badarg'),var('@c2')])]))])),clause([var('@c5'),var('@c4')],lit(atom,'true'),primop(lit(atom,'match_fail'),[tuple([lit(atom,'function_clause'),var('@c5'),var('@c4')])]))]))).

fundef(lit(atom,'mytrees'),var('join',2),fun([var('@c1'),var('@c0')],case([var('@c1'),var('@c0')],[clause([lit(atom,'leaf'),var('T2')],lit(atom,'true'),var('T2')),clause([tuple([lit(atom,'node'),var('X1'),var('L1'),var('R1')]),var('T2')],lit(atom,'true'),let([var('@c2')],apply(var('join',2),[var('L1'),var('R1')]),tuple([lit(atom,'node'),var('X1'),var('@c2'),var('T2')]))),clause([var('@c4'),var('@c3')],lit(atom,'true'),primop(lit(atom,'match_fail'),[tuple([lit(atom,'function_clause'),var('@c4'),var('@c3')])]))]))).

fundef(lit(atom,'mytrees'),var('delete',2),fun([var('@c1'),var('@c0')],case([var('@c1'),var('@c0')],[clause([var('@c6'),lit(atom,'leaf')],lit(atom,'true'),lit(atom,'leaf')),clause([var('X'),tuple([lit(atom,'node'),var('@c7'),var('L'),var('R')])],call(lit(atom,'erlang'),lit(atom,'=:='),[var('@c7'),var('X')]),apply(var('join',2),[var('L'),var('R')])),clause([var('X'),tuple([lit(atom,'node'),var('Y'),var('L'),var('R')])],lit(atom,'true'),let([var('@c3')],apply(var('delete',2),[var('X'),var('L')]),let([var('@c2')],apply(var('delete',2),[var('X'),var('R')]),tuple([lit(atom,'node'),var('Y'),var('@c3'),var('@c2')])))),clause([var('@c5'),var('@c4')],lit(atom,'true'),primop(lit(atom,'match_fail'),[tuple([lit(atom,'function_clause'),var('@c5'),var('@c4')])]))]))).

fundef(lit(atom,'mytrees'),var('tree_sum',1),fun([var('@c0')],case([var('@c0')],[clause([lit(atom,'leaf')],lit(atom,'true'),lit(int,0)),clause([tuple([lit(atom,'node'),var('X'),var('L'),var('R')])],lit(atom,'true'),let([var('@c2')],apply(var('tree_sum',1),[var('L')]),let([var('@c1')],apply(var('tree_sum',1),[var('R')]),let([var('@c3')],call(lit(atom,'erlang'),lit(atom,'+'),[var('@c2'),var('@c1')]),call(lit(atom,'erlang'),lit(atom,'+'),[var('@c3'),var('X')]))))),clause([var('@c4')],lit(atom,'true'),primop(lit(atom,'match_fail'),[tuple([lit(atom,'function_clause'),var('@c4')])]))]))).

fundef(lit(atom,'mytrees'),var('preorder',1),fun([var('@c0')],case([var('@c0')],[clause([lit(atom,'leaf')],lit(atom,'true'),nil),clause([tuple([lit(atom,'node'),var('X'),var('L'),var('R')])],lit(atom,'true'),let([var('@c2')],apply(var('preorder',1),[var('L')]),let([var('@c1')],apply(var('preorder',1),[var('R')]),let([var('@c3')],call(lit(atom,'erlang'),lit(atom,'++'),[var('@c2'),var('@c1')]),cons(var('X'),var('@c3')))))),clause([var('@c4')],lit(atom,'true'),primop(lit(atom,'match_fail'),[tuple([lit(atom,'function_clause'),var('@c4')])]))]))).

fundef(lit(atom,'mytrees'),var('tree1',1),fun([var('@c0')],let([var('@c2')],apply(var('tree1',1),[var('@c0')]),let([var('@c1')],apply(var('tree1',1),[var('@c0')]),call(lit(atom,'proper_types'),lit(atom,'union'),[cons(lit(atom,'leaf'),cons(tuple([lit(atom,'node'),var('@c0'),var('@c2'),var('@c1')]),nil))]))))).

fundef(lit(atom,'mytrees'),var('tree2',1),fun([var('@c0')],let([var('@c3')],fun([],let([var('@c2')],apply(var('tree2',1),[var('@c0')]),let([var('@c1')],apply(var('tree2',1),[var('@c0')]),tuple([lit(atom,'node'),var('@c0'),var('@c2'),var('@c1')])))),let([var('@c4')],call(lit(atom,'proper_types'),lit(atom,'lazy'),[var('@c3')]),call(lit(atom,'proper_types'),lit(atom,'union'),[cons(lit(atom,'leaf'),cons(var('@c4'),nil))]))))).

fundef(lit(atom,'mytrees'),var('tree3',1),fun([var('@c0')],let([var('@c3')],fun([var('@c1')],apply(var('tree3',2),[var('@c1'),var('@c0')])),call(lit(atom,'proper_types'),lit(atom,'sized'),[var('@c3')])))).

fundef(lit(atom,'mytrees'),var('tree3',2),fun([var('@c1'),var('@c0')],case([var('@c1'),var('@c0')],[clause([lit(int,0),var('@c11')],lit(atom,'true'),lit(atom,'leaf')),clause([var('S'),var('G')],lit(atom,'true'),let([var('@c2')],apply(var('tree3',2),[lit(int,0),var('G')]),let([var('@c7')],fun([],let([var('@c5')],call(lit(atom,'erlang'),lit(atom,'-'),[var('S'),lit(int,1)]),let([var('@c6')],apply(var('tree3',2),[var('@c5'),var('G')]),let([var('@c3')],call(lit(atom,'erlang'),lit(atom,'-'),[var('S'),lit(int,1)]),let([var('@c4')],apply(var('tree3',2),[var('@c3'),var('G')]),tuple([lit(atom,'node'),var('G'),var('@c6'),var('@c4')])))))),let([var('@c8')],call(lit(atom,'proper_types'),lit(atom,'lazy'),[var('@c7')]),call(lit(atom,'proper_types'),lit(atom,'union'),[cons(var('@c2'),cons(var('@c8'),nil))])))))]))).

fundef(lit(atom,'mytrees'),var('tree4',1),fun([var('@c0')],let([var('@c3')],fun([var('@c1')],apply(var('tree4',2),[var('@c1'),var('@c0')])),call(lit(atom,'proper_types'),lit(atom,'sized'),[var('@c3')])))).

fundef(lit(atom,'mytrees'),var('tree4',2),fun([var('@c1'),var('@c0')],case([var('@c1'),var('@c0')],[clause([lit(int,0),var('@c11')],lit(atom,'true'),lit(atom,'leaf')),clause([var('S'),var('G')],lit(atom,'true'),let([var('@c2')],apply(var('tree4',2),[lit(int,0),var('G')]),let([var('@c7')],fun([],let([var('@c5')],call(lit(atom,'erlang'),lit(atom,'-'),[var('S'),lit(int,1)]),let([var('@c6')],apply(var('tree4',2),[var('@c5'),var('G')]),let([var('@c3')],call(lit(atom,'erlang'),lit(atom,'-'),[var('S'),lit(int,1)]),let([var('@c4')],apply(var('tree4',2),[var('@c3'),var('G')]),tuple([lit(atom,'node'),var('G'),var('@c6'),var('@c4')])))))),let([var('@c8')],call(lit(atom,'proper_types'),lit(atom,'lazy'),[var('@c7')]),call(lit(atom,'proper_types'),lit(atom,'frequency'),[cons(tuple([lit(int,1),var('@c2')]),cons(tuple([lit(int,9),var('@c8')]),nil))])))))]))).

fundef(lit(atom,'mytrees'),var('tree5',1),fun([var('@c0')],let([var('@c3')],fun([var('@c1')],apply(var('tree5',2),[var('@c1'),var('@c0')])),call(lit(atom,'proper_types'),lit(atom,'sized'),[var('@c3')])))).

fundef(lit(atom,'mytrees'),var('tree5',2),fun([var('@c1'),var('@c0')],case([var('@c1'),var('@c0')],[clause([lit(int,0),var('@c11')],lit(atom,'true'),lit(atom,'leaf')),clause([var('S'),var('G')],lit(atom,'true'),let([var('@c2')],apply(var('tree5',2),[lit(int,0),var('G')]),let([var('@c7')],fun([],let([var('@c5')],call(lit(atom,'erlang'),lit(atom,'div'),[var('S'),lit(int,2)]),let([var('@c6')],apply(var('tree5',2),[var('@c5'),var('G')]),let([var('@c3')],call(lit(atom,'erlang'),lit(atom,'div'),[var('S'),lit(int,2)]),let([var('@c4')],apply(var('tree5',2),[var('@c3'),var('G')]),tuple([lit(atom,'node'),var('G'),var('@c6'),var('@c4')])))))),let([var('@c8')],call(lit(atom,'proper_types'),lit(atom,'lazy'),[var('@c7')]),call(lit(atom,'proper_types'),lit(atom,'frequency'),[cons(tuple([lit(int,1),var('@c2')]),cons(tuple([lit(int,9),var('@c8')]),nil))])))))]))).

fundef(lit(atom,'mytrees'),var('tree',1),fun([var('@c0')],let([var('@c3')],fun([var('@c1')],apply(var('tree',2),[var('@c1'),var('@c0')])),call(lit(atom,'proper_types'),lit(atom,'sized'),[var('@c3')])))).

fundef(lit(atom,'mytrees'),var('tree',2),fun([var('@c1'),var('@c0')],case([var('@c1'),var('@c0')],[clause([lit(int,0),var('@c14')],lit(atom,'true'),lit(atom,'leaf')),clause([var('S'),var('G')],lit(atom,'true'),let([var('@c2')],apply(var('tree',2),[lit(int,0),var('G')]),let([var('@c10')],fun([],let([var('@c6')],call(lit(atom,'erlang'),lit(atom,'div'),[var('S'),lit(int,2)]),let([var('@c7')],apply(var('tree',2),[var('@c6'),var('G')]),let([var('@c8')],call(lit(atom,'erlang'),lit(atom,'div'),[var('S'),lit(int,2)]),let([var('@c9')],apply(var('tree',2),[var('@c8'),var('G')]),let([var('@c5')],fun([var('@c3')],case([var('@c3')],[clause([cons(var('L'),cons(var('R'),nil))],lit(atom,'true'),tuple([lit(atom,'node'),var('G'),var('L'),var('R')])),clause([var('@c4')],lit(atom,'true'),primop(lit(atom,'match_fail'),[tuple([lit(atom,'function_clause'),var('@c4')])]))])),call(lit(atom,'proper_types'),lit(atom,'bind'),[cons(var('@c7'),cons(var('@c9'),nil)),var('@c5'),lit(atom,'true')]))))))),let([var('@c11')],call(lit(atom,'proper_types'),lit(atom,'lazy'),[var('@c10')]),call(lit(atom,'proper_types'),lit(atom,'frequency'),[cons(tuple([lit(int,1),var('@c2')]),cons(tuple([lit(int,9),var('@c11')]),nil))])))))]))).

spec(member(T,tree(T)),boolean).

spec(join(tree(T),tree(T)),tree(T)).

spec(delete(T,tree(T)),tree(T)).

spec(tree_sum(tree(number)),number).

spec(preorder(tree(T)),list(T)).

typedef(tree(T),exactly(lit(atom,leaf))).
typedef(tree(T),tuple([exactly(lit(atom,node)),T,tree(T),tree(T)])).

gen_prop_delete((X,T)) :-
  typeof(X,integer),
  typeof(T,tree(integer)),
  prop_eval(
    call(lit(atom,'erlang'),lit(atom,'not'),[apply(var('member',2),[var('X'),apply(var('delete',2),[var('X'),var('T')])])]),
    [('X',X),('T',T)],
    lit(atom,_Res)
  ).

gen_prop_sum((T)) :-
  typeof(T,tree(integer)),
  prop_eval(
    call(lit(atom,'erlang'),lit(atom,'=:='),[call(lit(atom,'lists'),lit(atom,'sum'),[apply(var('preorder',1),[var('T')])]),apply(var('tree_sum',1),[var('T')])]),
    [('T',T)],
    lit(atom,_Res)
  ).

